rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,223 @@
+
+#### Check: Smoke tests pass (i.e. the app runs in a production environment)
+
+For the purpose of automating deployments:
+
+- We need to test that an app can respond to a request for each of the machine classes it runs on [[1](https://github.com/alphagov/govuk-app-deployment/blob/d4bb62aa79face46d2e164e316f38f7a64a1b8be/content-store/config/deploy.rb#L3)] [[2](https://github.com/alphagov/govuk-app-deployment/blob/d4bb62aa79face46d2e164e316f38f7a64a1b8be/whitehall/config/deploy.rb#L13)]. This is purely to prove that an instance of the app is running [[1](https://github.com/alphagov/email-alert-api/pull/1346)], and may be covered by other tests.
+
+- We need to test that an app can connect to any systems it uses for data storage and retrieval, such as a database. This usually has production-specific config [[1](https://github.com/alphagov/email-alert-api/blob/1b5f59f3927637afb0c790702f8f0050788edc9d/config/database.yml#L22)] [[2](https://github.com/alphagov/content-store/blob/f4b61d4f2756341b6418085d878ec83b09d70e9c/config/mongoid.yml#L23)] [[3](https://github.com/alphagov/content-publisher/blob/944fe51b4cb8205348e155b965b8bad57adcf4fe/config/storage.yml#L5)], which could be faulty.
+
+- We do not need to test connectivity to other GOV.UK apps, since this should be handled by [our internal API adapters](https://github.com/alphagov/gds-api-adapters/blob/1ceba690edb37295257328d953874224833fcd5e/lib/gds_api.rb) with no configuration inside the repo. An exception to this is [authentication with Signon](https://github.com/alphagov/email-alert-api/blob/1b5f59f3927637afb0c790702f8f0050788edc9d/config/initializers/gds_sso.rb).
+
+With this in mind, an app has "enough" of these tests when:
+
+- It has a test that checks for a successful response to an arbitrary request.
+- It has a test that checks for a successful login with Signon (if applicable).
+- It has a test that checks for connectivity to any read/write dependencies:
+
+  - Databases e.g. MongoDB, Postgres, MySQL, Redis
+  - Queueing systems e.g. RabbitMQ
+  - File systems e.g. AWS S3
+  - Caching systems e.g. Memcached
+
+> Example: [Travel Advice Publisher](https://github.com/alphagov/travel-advice-publisher) has enough of these tests because it has one to check login with Signon [[1](https://github.com/alphagov/smokey/blob/a60df2cde5c886fa6d54e8cd820d90facca89e8e/features/publishing_tools.feature#L106)] and another to check its `/healthcheck` endpoint for connectivity to Redis and MongoDB [[1](https://github.com/alphagov/smokey/blob/a60df2cde5c886fa6d54e8cd820d90facca89e8e/features/travel_advice_publisher.feature)] [[2](https://github.com/alphagov/travel-advice-publisher/pull/979)].
```

Comment:
I'll just respond to the networking issue first: I don't think that will be a big issue, since the monitoring machine Smokey currently runs on has pretty broad access (for Icinga) to all of our machines.

Local network tests [aren't new with CD](https://github.com/alphagov/smokey/blob/160fbe09541ed5d1d0acbdf44355ec5d454c0dc1/features/mirror.feature). However, I agree it's not ideal to run healthcheck tests as part of [the Smokey loop](https://docs.publishing.service.gov.uk/manual/alerts/smokey-loop-tests.html), when they were only added for the other, manual use case for Smokey. 

(For anyone else reading: we can't easily check Icinga's healthcheck alerts, since they're per-machine and the checks are asynchronous, so we'd have to wait a while to be sure they'd refreshed.)

With [the recent work to restart our apps reliably](https://github.com/alphagov/govuk-puppet/pull/10761), it would be much easier to introduce a healthcheck on the end of the existing "Deploy_App" job. On balance, I don't think it does much harm to have healthcheck tests in Smokey, but it would be much more DRY if the "Deploy_App" job could do them itself. For the sake of unblocking CD, I'd like to continue as-is, but add a new consequence to see if we can do better than what currently exists:

### Healthy deployments

The original implementation of the CD pipeline resulted in [a new set of `/healthcheck` tests in Smokey](https://github.com/alphagov/smokey/pull/731), in addition to checking them with an Icinga alert on each machine. It's still impractical to use Icinga as part of the CD pipeline: the alerts are checked asynchronously and the names are unstable. Some duplication is therefore inevitable.

An alternative to individual Smokey tests is to make calling the `/healthcheck` endpoint part of the "Deploy_App" job, such that it would fail if the healthcheck fails. We will investigate this approach as a way to DRY up the Smokey tests, in parallel with enabling automatic deployments for our apps.

Reply:
Sure, that sounds quite reasonable. I didn't realise about the mirror tests (there's always one small inconsistency from the past that ends up breeding in the future)

It might well be that the need for the healthcheck Smokey is rendered moot by replatforming - as it would seem a future deployment approach should consider an application healthy (via the healthcheck) before switching requests to it.